substratevm/src/com.oracle.svm.hosted.jdk11/src/com/oracle/svm/hosted/ModuleLayerFeature.java

```diff
@@ -119,49 +139,183 @@ public void afterAnalysis(AfterAnalysisAccess access) {
+                    moduleExportedPackagesField.set(m, exportedPackages);
+                } else {
+                    Map<String, Set<Module>> openPackages = new HashMap<>();
+                    Map<String, Set<Module>> exportedPackages = new HashMap<>();
+                    for (ModuleDescriptor.Opens opens : descriptor.opens()) {
+                        String source = opens.source();
+                        if (opens.isQualified()) {
+                            Set<Module> targets = new HashSet<>();
+                            for (String target : opens.targets()) {
+                                Module m2 = (Module) moduleFindModuleMethod.invoke(null, target, Map.of(), nameToModule, runtimeBootLayer.parents());
+                                if (m2 != null) {
+                                    targets.add(m2);
+                                }
+                            }
+                            if (!targets.isEmpty()) {
+                                openPackages.put(source, targets);
+                            }
+                        } else {
+                            openPackages.put(source, everyoneSet);
+                        }
+                    }
+
+                    for (ModuleDescriptor.Exports exports : descriptor.exports()) {
+                        String source = exports.source();
+                        Set<Module> openToTargets = openPackages.get(source);
+                        if (openToTargets != null && openToTargets.contains(everyoneModule)) {
+                            continue;
+                        }
+
+                        if (exports.isQualified()) {
+                            Set<Module> targets = new HashSet<>();
+                            for (String target : exports.targets()) {
+                                Module m2 = (Module) moduleFindModuleMethod.invoke(null, target, Map.of(), nameToModule, runtimeBootLayer.parents());
+                                if (m2 != null) {
+                                    if (openToTargets == null || !openToTargets.contains(m2)) {
+                                        targets.add(m2);
+                                    }
+                                }
+                            }
+                            if (!targets.isEmpty()) {
+                                exportedPackages.put(source, targets);
+                            }
+                        } else {
+                            exportedPackages.put(source, everyoneSet);
+                        }
+                    }
+
+                    moduleOpenPackagesField.set(m, openPackages);
+                    moduleExportedPackagesField.set(m, exportedPackages);
+                }
+            }
+        }
+
+        return nameToModule;
+    }
+
+    /*
+     * Creates a stream of module names that are reachable from a given module through "requires"
+     */
+    private static Stream<String> extractRequiredModuleNames(Module m) {
```

Comment:
Maybe move next to caller (right after `afterAnalysis`)

Reply:
Done.